### PR TITLE
🌱 test: add dynamic finalizer assertions for e2e framework

### PR DESCRIPTION
**Pull request by adityabhatia**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubernetes-sigs/cluster-api/issues/9969

/area testing

### CODE REVIEW

**Comment by sbueringer** [comment]
@adityabhatia There seems to be something wrong with the commit. The diff looks correct in GitHub but based on the commit title and locally it is a merged commit.

File: test/framework/finalizers_helpers.go [hunk 1]
```diff
@@ -111,6 +128,8 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 		setFinalizers := obj.GetFinalizers()
 
 		if len(setFinalizers) > 0 {
+			// assert if the expected finalizers are set on the resource
+			Expect(allFinalizerAssertions[node.Object.Kind]).To(Equal(setFinalizers), "for resource type ", node.Object.Kind)
```
Comment (sbueringer) [inline test/framework/finalizers_helpers.go:132]: first parameter should be actual

File: test/framework/finalizers_helpers.go [hunk 2]
```diff
@@ -149,6 +168,18 @@ func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace st
 	}).WithTimeout(1 * time.Minute).WithPolling(2 * time.Second).Should(Succeed())
 }
 
+// concatenateFinalizerAssertions concatenates all finalizer assertions into one map.
+func concatenateFinalizerAssertions(finalizerAssertions ...map[string][]string) map[string][]string {
+	allFinalizerAssertions := make(map[string][]string, 0)
+	for i := range finalizerAssertions {
+		for kind, finalizers := range finalizerAssertions[i] {
+			allFinalizerAssertions[kind] = finalizers
```
Comment (sbueringer) [inline test/framework/finalizers_helpers.go:176]: Maybe we should throw an error if we find duplicate assertions for the same kind

Comment (adityabhatia) [reply test/framework/finalizers_helpers.go]: True, I added aggregated errors for such cases.

File: test/framework/finalizers_helpers.go [hunk 3]
```diff
@@ -106,6 +124,8 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 		setFinalizers := obj.GetFinalizers()
 
 		if len(setFinalizers) > 0 {
+			// assert if the expected finalizers are set on the resource
+			Expect(setFinalizers).To(Equal(allFinalizerAssertions[node.Object.Kind]), "for resource type ", node.Object.Kind)
```
Comment (sbueringer) [inline test/framework/finalizers_helpers.go:128]:
```suggestion
			Expect(setFinalizers).To(Equal(allFinalizerAssertions[node.Object.Kind]), "for resource type %s", node.Object.Kind)
```

optionalDescription is handled like this for len >= 2
```go
return fmt.Sprintf(optionalDescription[0].(string), optionalDescription[1:]...) + "\n"
```

File: test/framework/finalizers_helpers.go [hunk 4]
```diff
@@ -143,3 +163,24 @@ func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace st
...
+
+// concatenateFinalizerAssertions concatenates all finalizer assertions into one map. It reports errors if assertions already exist.
+func concatenateFinalizerAssertions(finalizerAssertions ...map[string][]string) (map[string][]string, error) {
+	var allErrs []error
+	allFinalizerAssertions := make(map[string][]string, 0)
+
+	for i := range finalizerAssertions {
+		for kind, finalizers := range finalizerAssertions[i] {
+			if _, alreadyExists := allFinalizerAssertions[kind]; alreadyExists {
+				allErrs = append(allErrs, fmt.Errorf("finaliser assertion cannot be applied as it already exists for kind: %s, existing value: %v, new value: %v",
```
Comment (sbueringer) [inline test/framework/finalizers_helpers.go:175]:
```suggestion
				allErrs = append(allErrs, fmt.Errorf("finalizer assertion cannot be applied as it already exists for kind: %s, existing value: %v, new value: %v",
```

**Comment by sbueringer** [comment]
@adityabhatia Happy to merge this PR now that we reverted the other PR that broke CI (+/- addressing nits / rebase). To unblock the work in CAPV>

File: test/framework/finalizers_helpers.go (hunk 3 above)
Comment (adityabhatia) [reply test/framework/finalizers_helpers.go]: Always thought optional arguments are concatenated

**Comment by adityabhatia** [comment]
> @adityabhatia Happy to merge this PR now that we reverted the other PR that broke CI (+/- addressing nits / rebase)., to unblock the work in CAPV.

Rebased on main (post revert) and ready for review :)

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**